overlord/devicestate/devicestate_recovery_keys_test.go

```diff
@@ -0,0 +1,174 @@
+}
+
+func mockSystemRecoveryKeys(c *C, alsoReinstall bool) {
+	// same inputs/outputs as secboot:crypt_test.go in this test
+	rkeystr, err := hex.DecodeString("e1f01302c5d43726a9b85b4a8d9c7f6e")
+	c.Assert(err, IsNil)
+	mockSnapFDEFile(c, "recovery.key", []byte(rkeystr))
+
+	if alsoReinstall {
+		skeystr := "1234567890123456"
+		mockSnapFDEFile(c, "reinstall.key", []byte(skeystr))
+	}
+}
+
+func (s *deviceMgrRecoveryKeysSuite) TestEnsureRecoveryKeysBackwardCompat(c *C) {
+	mockSystemRecoveryKeys(c, true)
+
+	keys, err := s.mgr.EnsureRecoveryKeys()
+	c.Assert(err, IsNil)
+
+	c.Assert(keys, DeepEquals, &client.SystemRecoveryKeysResponse{
+		RecoveryKey:  "61665-00531-54469-09783-47273-19035-40077-28287",
+		ReinstallKey: "12849-13363-13877-14391-12345-12849-13363-13877",
+	})
+}
+
+func (s *deviceMgrRecoveryKeysSuite) TestEnsureRecoveryKey(c *C) {
+	_, err := s.mgr.EnsureRecoveryKeys()
+	c.Check(err, ErrorMatches, `system does not use disk encryption`)
+
+	rkeystr, err := hex.DecodeString("e1f01302c5d43726a9b85b4a8d9c7f6e")
+	c.Assert(err, IsNil)
+	defer devicestate.MockSecbootEnsureRecoveryKey(func(keyFile string, mountPoints []string) (keys.RecoveryKey, error) {
+		c.Check(keyFile, Equals, filepath.Join(dirs.SnapFDEDir, "recovery.key"))
+		c.Check(mountPoints, DeepEquals, []string{boot.InitramfsDataDir, boot.InitramfsUbuntuSaveDir})
+
+		var rkey keys.RecoveryKey
+		copy(rkey[:], []byte(rkeystr))
+		return rkey, nil
+	})()
+	mockSnapFDEFile(c, "marker", nil)
+
+	keys, err := s.mgr.EnsureRecoveryKeys()
+	c.Assert(err, IsNil)
+
+	c.Assert(keys, DeepEquals, &client.SystemRecoveryKeysResponse{
+		RecoveryKey: "61665-00531-54469-09783-47273-19035-40077-28287",
+	})
+}
+
+func (s *deviceMgrRecoveryKeysSuite) TestEnsureRecoveryKeyInstallMode(c *C) {
+	devicestate.SetSystemMode(s.mgr, "install")
+
+	rkeystr, err := hex.DecodeString("e1f01302c5d43726a9b85b4a8d9c7f6e")
+	c.Assert(err, IsNil)
+	defer devicestate.MockSecbootEnsureRecoveryKey(func(keyFile string, mountPoints []string) (keys.RecoveryKey, error) {
+		c.Check(keyFile, Equals, filepath.Join(boot.InstallHostFDEDataDir, "recovery.key"))
+		c.Check(mountPoints, DeepEquals, []string{filepath.Dir(boot.InstallHostWritableDir), boot.InitramfsUbuntuSaveDir})
+
+		var rkey keys.RecoveryKey
+		copy(rkey[:], []byte(rkeystr))
+		return rkey, nil
+	})()
+
+	p := filepath.Join(boot.InstallHostFDEDataDir, "marker")
+	err = os.MkdirAll(filepath.Dir(p), 0755)
+	c.Assert(err, IsNil)
+	err = ioutil.WriteFile(p, nil, 0644)
+	c.Assert(err, IsNil)
+
+	keys, err := s.mgr.EnsureRecoveryKeys()
+	c.Assert(err, IsNil)
+
+	c.Assert(keys, DeepEquals, &client.SystemRecoveryKeysResponse{
+		RecoveryKey: "61665-00531-54469-09783-47273-19035-40077-28287",
+	})
+}
+
+func (s *deviceMgrRecoveryKeysSuite) TestRemoveRecoveryKeys(c *C) {
+	err := s.mgr.RemoveRecoveryKeys()
+	c.Check(err, ErrorMatches, `system does not use disk encryption`)
+
+	called := false
+	rkey := filepath.Join(dirs.SnapFDEDir, "recovery.key")
```

Comment:
GitHub actions thinks that `rkey` is unused, but it's probably because it doesn't realize that it's captured by the lambda?

Reply:
I need to double check